File: haskell/rotational-cipher/src/RotationalCipher.hs
module RotationalCipher
  ( rotate,
  )
where

import Data.Char (chr, isLower, isUpper, ord)

-- | Rotate a string's letters by a given number of letters.
--
-- > rotate 5 "omg" == "trl"
-- > rotate 0 "c" == "c"
-- > rotate 26 "Cool" == "Cool"
-- > rotate 13 "The quick brown fox jumps over the lazy dog." == "Gur dhvpx oebja sbk whzcf bire gur ynml qbt."
rotate :: Int -> String -> String
rotate n
  | n `mod` 26 == 0 = id
  | otherwise = map (shift n)

-- | Shift a letter by a specified number of letters.
--
-- > shift 3 'a' == 'd'
-- > shift 2 'z' == 'b'
-- > shift 8 'E' == 'M'
-- > shift 13 '.' == '.'
shift :: Int -> Char -> Char
shift n c
  | isLower c = shift' 97 n c -- ord 'a' == 97
  | isUpper c = shift' 65 n c -- ord 'A' == 65
  | otherwise = c

-- | Shift a letter by a specified number of letters, wrapping around a given
-- 'Int' value for the letter A, i.e. @97@ or @65@.
--
-- Use @shift' 97@ for lowercase and @shift' 65@
-- for uppercase letters, since @ord ''a'' == 97@ and @ord ''A'' == 65@.
shift' :: Int -> Int -> Char -> Char
shift' ordA n c = chr (ordA + ((ord c - ordA + n) `mod` 26))

File: julia/rotational-cipher/src/RotationalCipher.jl
module RotationalCipher

export rotate


for i in range(0, 26)
    @eval export $(Symbol("@R$(i)_str"))
    @eval macro $(Symbol("R$(i)_str"))(s)
        rotate($i, s)
    end
end


"""
    rotate(n::Integer, input::AbstractString)

Rotate a string `input` by `n`.

```jldoctest
julia> R5"omg"
"trl"

julia> R0"c"
"c"

julia> R26"Cool"
"Cool"

julia> R13"The quick brown fox jumps over the lazy dog."
"Gur dhvpx oebja sbk whzcf bire gur ynml qbt."

julia> R13"Gur dhvpx oebja sbk whzcf bire gur ynml qbt."
"The quick brown fox jumps over the lazy dog."
```
"""
function rotate(n::Integer, input::AbstractString)::String
    if n % 26 ≡ 0
        return input
    end

    map(char -> rotate(n, char), input)
end


"""
    rotate(n::Integer, char::AbstractChar)::Char

Rotate an ASCII letter by `n`.
"""
function rotate(n::Integer, char::AbstractChar)::Char
    if !(isascii(char) && isletter(char))
        return char
    end

    pivot = islowercase(char) ? 'a' : 'A'
    pivot + (char - pivot + n) % 26
end

end # module RotationalCipher
